boa/src/builtins/json/mod.rs

```diff
@@ -62,10 +67,62 @@ pub fn parse(_: &mut Value, args: &[Value], _: &mut Interpreter) -> ResultValue
-pub fn stringify(_: &mut Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
-    let obj = args.get(0).expect("cannot get argument for JSON.stringify");
-    let json = obj.to_json().to_string();
-    Ok(Value::from(json))
+pub fn stringify(_: &mut Value, args: &[Value], interpreter: &mut Interpreter) -> ResultValue {
+    let object = args.get(0).expect("cannot get argument for JSON.stringify");
+    if args.len() == 1 {
+        let json = object.to_json().to_string();
+        return Ok(Value::from(json));
+    }
+
+    let object_to_return = Value::new_object(None);
+    if let Some(arg) = args.get(1) {
+        if let ValueData::Object(ref obj) = arg.data() {
+            let replacer = (*obj).deref().borrow();
+            match replacer.kind {
+                ObjectKind::Array => {
+                    for (key, value) in replacer.properties.iter() {
+                        if key == "length" {
+                            continue;
+                        }
+
+                        if let Some(Value(x)) = &value.value {
+                            object_to_return.set_property(
+                                x.to_string(),
+                                object
+                                    .get_property(&x.to_string())
+                                    .expect("failed to get property from object"),
+                            );
+                        }
+                    }
+                }
+                ObjectKind::Function => {
+                    if let ValueData::Object(ref obj) = object.data() {
+                        let object_to_stringify = (*obj).deref().borrow();
+                        for (key, val) in object_to_stringify.properties.iter() {
+                            if let Some(value) = &val.value {
+                                let mut this_arg = object.clone();
+                                object_to_return.set_property(
+                                    String::from(key),
+                                    Property::default().value(
+                                        interpreter
+                                            .call(
+                                                arg,
+                                                &mut this_arg,
+                                                &[Value::string(key), Value::from(value)],
+                                            )
+                                            .expect("failed to get returned value from replacer function"),
+                                    ),
+                                );
+                            }
+                        }
+                    }
+                }
+                _ => panic!("JSON.stringify replacer can only be an array or function"),
+            }
+            return Ok(Value::from(object_to_return.to_json().to_string()));
+        }
+    }
+    panic!("cannot get replacer for JSON.stringify");
```

Comment:
I'm really not in love with the perceived complexity of this function. Any suggestions on making it more succinct would be greatly appreciated, but I'm pretty new to Rust.

Reply:
Let me check if this can be improved :)